trl/trainer/orpo_trainer.py

```diff
@@ -59,6 +59,212 @@
+        if len(p) != len(m):
+            raise ValueError("Prompt input ids and attention mask should have the same length.")
+
+    chosen_input_ids = [f[r:] for f, r in zip(full_input_ids, response_token_ids_start_idx)]
+    chosen_attention_mask = [f[r:] for f, r in zip(prompt_and_chosen_attention_mask, response_token_ids_start_idx)]
+
+    # prompt and rejected now
+    prompt_and_rejected = [prompt + rejected for prompt, rejected in zip(batch["prompt"], batch["rejected"])]
+    prompt_and_rejected_tokenized = tokenizer(prompt_and_rejected, add_special_tokens=False)
+    prompt_and_rejected_input_ids = prompt_and_rejected_tokenized["input_ids"]
+    prompt_and_rejected_attention_mask = prompt_and_rejected_tokenized["attention_mask"]
+
+    rejected_input_ids = [f[len(p) :] for f, p in zip(prompt_and_rejected_input_ids, prompt_input_ids)]
+    rejected_attention_mask = [f[len(p) :] for f, p in zip(prompt_and_rejected_attention_mask, prompt_attention_mask)]
+
+    # Concat tokens to form `enc(a) + enc(a + b)[len(enc(a)):]`
+    full_concat_input_ids = [np.concatenate([p, a]) for p, a in zip(prompt_input_ids, rejected_input_ids)]
+    # Prepare input tokens for token by token comparison
+    full_input_ids = [np.array(f) for f in prompt_and_rejected_input_ids]
+    for full, concat in zip(full_input_ids, full_concat_input_ids):
+        if len(full) != len(concat):
+            raise ValueError(
+                f"Prompt input ids and chosen input ids should have the same length. Found {len(full)} != {len(concat)}"
+            )
+
+    # On some tokenizers, like Llama-2 tokenizer, there are occasions where tokens
+    # can be merged together when tokenizing prompt+answer. This could result
+    # on the last token from the prompt being different when tokenized on its own
+    # vs when done as prompt+answer.
+    response_token_ids_start_idx = [len(p) for p in prompt_input_ids]
+
+    # If tokenized prompt is different than both prompt+answer, then it means the
+    # last token has changed due to merging.
+    for idx, (p, f, r) in enumerate(zip(prompt_input_ids, full_input_ids, response_token_ids_start_idx)):
+        if not np.array_equal(p, f[:r]):
+            response_token_ids_start_idx[idx] -= 1
+
+    prompt_input_ids = [f[:r] for f, r in zip(full_input_ids, response_token_ids_start_idx)]
+    prompt_attention_mask = [f[:r] for f, r in zip(prompt_and_rejected_attention_mask, response_token_ids_start_idx)]
+
+    for p, m in zip(prompt_input_ids, prompt_attention_mask):
+        if len(p) != len(m):
+            raise ValueError("Prompt input ids and attention mask should have the same length.")
+
+    rejected_input_ids = [f[r:] for f, r in zip(full_input_ids, response_token_ids_start_idx)]
+    rejected_attention_mask = [f[r:] for f, r in zip(prompt_and_rejected_attention_mask, response_token_ids_start_idx)]
+
+    return dict(
+        prompt_input_ids=prompt_input_ids,
```

Comment:
Note there is one _small_ difference with the new implementation: we do not distinguish between the prompt input IDs of the chosen / rejected pairs, favouring instead to have a single set of IDs per prompt.

In general, I think the processing logic could be simplified significantly, but for now I've focused on ensuring parity with the current version